File: 5. Courses/1. CS50 Intro To CS/lec2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
int lettersNumbers(string sentence);
int wordsNumbers(string sentence);
int sentencesNumbers(string sentence);
int main()
{
    string sentence = get_string("Text: ");
    int l = lettersNumbers(sentence);
    int s = sentencesNumbers(sentence);
    int w = wordsNumbers(sentence);
    // Average no. of letters:
    float aS = (float) s / w * 100;
    float aL = (float) l / w * 100;
    // Formula to calculate the level of english:
    float index = 0.0588 * aL - 0.296 * aS - 15.8;
    int index1 = round(index);

    if (index1 < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index1 >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index1);
    }
}
// A Function That Takes The string and give back the number of words:
int lettersNumbers(string sentence)
{
    int i = 0;
    int j = 0;
    while (sentence[i] != '\0')
    {
        if (isalpha(sentence[i]))
        {
            j++;
        }
        i++;
    }
    return j;
}
// A Function That Takes The string and give back the number of words:
int wordsNumbers(string sentence)
{
    int i = 0;
    int j = 0;
    while (sentence[i] != '\0')
    {
        if (sentence[i] == ' ')
        {
            j++;
        }
        i++;
    }
    // because there are is one less space than words:
    j++;
    return j;
}
// A Function That Takes The string and give back the number of Sentences:
int sentencesNumbers(string sentence)
{
    int i = 0;
    int j = 0;
    while (sentence[i] != '\0')
    {
        if (sentence[i] == '.' || sentence[i] == '!' || sentence[i] == '?')
        {
            j++;
        }
        i++;
    }
    return j;
}

File: 5. Courses/1. CS50X/tests/lec6/sentimental-readability/readability.py
from cs50 import get_string

# Get the text input from the user
text = get_string("Text: ")

# Number of letters in the text (ignoring punctuation, spaces, and digits)
letters = 0
for c in text:
    if c.isalpha():
        letters += 1

# Number of words in the text (split by spaces)
words = 0
for word in text.split():
    if any(c.isalpha() for c in word):  # Ensure the word contains at least one letter
        words += 1

# Number of sentences in the text (count '.', '!', and '?')
sentences = text.count('.') + text.count('!') + text.count('?')

# Check if no words were found to avoid division by zero
if words == 0:
    print("No words found.")
else:
    # Average number of letters per 100 words
    L = letters / words * 100

    # Average number of sentences per 100 words
    S = sentences / words * 100

    # Coleman-Liau index formula
    grade = 0.0588 * L - 0.296 * S - 15.8

    # Output the grade based on the calculated index
    if grade < 1:
        print("Before Grade 1")
    elif grade >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(grade)}")
